File: leetcode/py/21.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
	    d = ListNode()
	    tail = d

        while list1 != None and list2 != None:
            if list1.val >= list2.val:
                tail.next = ListNode(val=list2.val)
                list2 = list2.next
        else:
            tail.next = ListNode(val=list1.val)
            list1 = list1.next

        if list1:
            tail.next = list1
        else:
            tail.next = list2

        return d.next

File: leetcode/go/21.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func setDefault() ListNode {
	ln := ListNode{}
	ln.Val = 0
	return ln
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {

	dummy := setDefault()
	tail := dummy

	for list1.Next != nil && list2.Next != nil {
		if list1.Val >= list2.Val {
			tail.Next = list2
			list2.Val = list2.Next.Val
			list2.Next = list2.Next.Next
		} else {
			tail.Next = list1
			list1.Val = list1.Next.Val
			list1.Next = list1.Next.Next
		}
	}

	if list1.Next == nil {
		for list2.Next != nil {
			tail.Next = list2
			list2.Val = list2.Next.Val
			list2.Next = list2.Next.Next
		}
	} else {
		for list1.Next != nil {
			tail.Next = list1
			list1.Val = list1.Next.Val
			list1.Next = list1.Next.Next
		}
	}

	return dummy.Next
}

func main() {
	a := setDefault()
	b := setDefault()

	fmt.Println(mergeTwoLists(a, b))

}
